connector/src/main/java/org/jboss/as/connector/subsystems/resourceadapters/AdminObjectAdd.java

```diff
@@ -55,11 +66,23 @@ protected void populateModel(ModelNode operation, ModelNode modelNode) throws Op
     }
 
     @Override
-    protected void performRuntime(OperationContext context, ModelNode operation, ModelNode recoveryEnvModel) throws OperationFailedException {
+    protected void performRuntime(OperationContext context, ModelNode operation, final Resource resource) throws OperationFailedException {
 
         final ModelNode address = operation.require(OP_ADDR);
         PathAddress path = PathAddress.pathAddress(address);
-        final String archiveName = path.getElement(path.size() - 2).getValue();
+        final String raName = path.getElement(path.size() - 2).getValue();
+        final String archiveOrModuleName;
+        ModelNode raModel = context.readResourceFromRoot(path.subAddress(0, path.size() - 1)).getModel();
+        final boolean statsEnabled = STATISTICS_ENABLED.resolveModelAttribute(context, raModel).asBoolean();
+
+        if (!raModel.hasDefined(ARCHIVE.getName()) && !raModel.hasDefined(MODULE.getName())) {
+            throw ConnectorLogger.ROOT_LOGGER.archiveOrModuleRequired();
+        }
+        if (raModel.get(ARCHIVE.getName()).isDefined()) {
+            archiveOrModuleName = raModel.get(ARCHIVE.getName()).asString();
```

Comment:
never resolve model like this.
you should do archiveOrModuleName = ARCHIVE.resolveModelAttribute(context, raModel).asString()
this will also enforce few things and take default value in mind.